WordPress/Classes/Utility/ReaderPostCacheProvider.swift

```diff
@@ -0,0 +1,74 @@
+import Foundation
+
+@objc class ReaderPostCacheProvider: NSObject {
+
+    let context: NSManagedObjectContext
+    init(context: NSManagedObjectContext) {
+        self.context = context
+    }
+
+    /**
+     Gets an stored ReaderPost from the local database, it searches for an Original Post,
+     a SourceAttribution Post or a Cross Post.
+
+     - Parameters:
+        - postID: the postID of the post to fetch
+        - siteID: the siteID of the post to fetch, can be a blogID if it has source attribution information
+
+     - Returns: A ReaderPost object if it is cached localy or nil if not found
+     */
+    func getPostByID(postID: Int, siteID: Int) -> ReaderPost? {
+
+        let fetchRequest = searchPostRequestForID(postID, siteID: siteID)
+        var post: ReaderPost? = nil
+
+        do {
+
+            let results = try context.executeFetchRequest(fetchRequest)
+            post = suitablePostFromPosts(results)
+
+        } catch {
+            DDLogSwift.logError("Error fetching post from database")
+        }
+
+        return post
+    }
+
+    // This functions checks from a multiple cachedPosts the original one and returns it
+    // If it can't find one, then returns the first one.
+    private func suitablePostFromPosts(posts: [AnyObject]?) -> ReaderPost? {
+
+        guard let cachedPosts = posts as? [ReaderPost] else {
+            return nil
+        }
+
+        let originalPosts = cachedPosts.filter { $0.sourceAttribution == nil }
+        return originalPosts.first ?? cachedPosts.first
```

Comment:
Thinking about this... 
While offline, if a user is viewing the Discover stream and taps a post to view its detail I'm imagining the postID and siteID specified in the source attribution will be passed to `getPostById:siteID:` to find a cached post, yah?  Wouldn't this end up returning the post already visible in the Discover stream if the `originalPost` wasn't already cached?
Is that intended or am I missing something?